stan/math/prim/mat/fun/gp_matern32_cov.hpp

```diff
@@ -0,0 +1,304 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_GP_MATERN32_COV_HPP
+#define STAN_MATH_PRIM_MAT_FUN_GP_MATERN32_COV_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/mat/fun/divide_columns.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive_finite.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <stan/math/prim/scal/fun/divide.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/meta/size_of.hpp>
+#include <cmath>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Returns a Matern 3/2 covariance matrix
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \frac{\sqrt{3}d(x, x')}{l})exp(-\frac{\sqrt{3}d(x, x')}{l})
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type for each scalar
+ * @tparam T_s type of parameter of sigma
+ * @tparam T_l type of parameter length scale
+ *
+ * @param x std::vector of scalars that can be used in squared distance
+ * @param length_scale length scale
+ * @param sigma marginal standard deviation or magnitude
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern32_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const T_l &length_scale) {
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = size_of(x);
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  const char *function = "gp_matern32_cov";
+  for (size_t i = 0; i < x_size; ++i)
```

Comment:
This logic doesn't seem right. It'll do the right thing, but it's really wasteful. `x_size * (x_size - 1) / 2` times, but you should only need to make `x_size` comparisons.

So, first off, it's perfectly acceptable to document this to say `@param x std::vector of scalars that can be used in square distance. It's assumed each element of x is of the same size`. Then we could just check the first element. If you don't want to do that, the logic here should be something like:
```
size_t x_element_size = size_of(x[0]);
for (size_t i = 1; i < x_size; ++i)
  check_size_match(function, "x row", size_of(x[i]), "x's other row", x_element_size);
```

Since we've already checked that `x_size > 0` (line 51), this is ok to do this way.